src/check-dsl.ts

```diff
@@ -27,109 +41,144 @@ const defaultError = (lines: any) => {
   };
 };
 
-export const checkDSL = (codeInEditor: string) => {
-  const lines = codeInEditor.split("\n");
-  const markers: any = [];
-  const reporter = report({ lines, markers });
+// helper function to populate the relationsPerType and globalRelations list
+// will throw error if there are duplication relations
+const populateRelation = (
+  lines: string[],
+  reporter: any,
+  parserResults: ParserResult,
+  globalRelations: Record<string, boolean>,
+  transformedTypes: Record<string, TransformedType>,
+) => {
+  // Looking at the types
+  parserResults.types.forEach((typeDef) => {
+    const typeName = typeDef.type;
+
+    // check to see if the typeName is valid
+    if (typeName === Keywords.SELF || typeName === ReservedKeywords.THIS) {
+      const lineIndex = getTypeLineNumber(typeName, lines);
+      reporter.reservedType({ lineIndex, value: typeName });
+    }
 
-  try {
-    const parserResults = parseDSL(codeInEditor);
-    const relationsPerType: Record<string, Record<string, boolean>> = {};
-    const globalRelations: Record<string, boolean> = { [Keywords.SELF]: true };
+    // Include keyword
+    const encounteredRelationsInType: Record<string, boolean> = { [Keywords.SELF]: true };
+    const relations: Record<string, RelationDefParserResult<RelationDefOperator>> = {};
 
-    // Looking at the types
-    parserResults.types.forEach((typeDef) => {
-      const typeName = typeDef.type;
+    typeDef.relations.forEach((relationDef) => {
+      const { relation: relationName } = relationDef;
+      relations[relationName] = relationDef;
 
-      // Include keyword
-      const encounteredRelationsInType: Record<string, boolean> = { [Keywords.SELF]: true };
+      if (relationName === Keywords.SELF || relationName === ReservedKeywords.THIS) {
+        const lineIndex = getRelationLineNumber(relationName, lines);
+        reporter.reservedRelation({ lineIndex, value: relationName });
+      } else if (encounteredRelationsInType[relationName]) {
+        // Check if we have any duplicate relations
+        // figure out what is the lineIdx in question
+        const initialLineIdx = getRelationLineNumber(relationName, lines);
+        const duplicateLineIdx = getRelationLineNumber(relationName, lines, initialLineIdx + 1);
+        reporter.duplicateDefinition({ lineIndex: duplicateLineIdx, value: relationName });
+      }
 
-      typeDef.relations.forEach((relationDef) => {
-        const { relation: relationName } = relationDef;
+      encounteredRelationsInType[relationName] = true;
+      globalRelations[relationName] = true;
+    });
+    transformedTypes[typeName] = {
+      comment: "",
+      type: typeName,
+      relations,
+    };
+  });
```

Comment:
```suggestion
  return [globalRelations, transformedTypes] as const
  });
  ```
  and get rid of the function parameters?

Reply:
done